src/experiment/evaluation.jl

```diff
@@ -8,40 +23,67 @@ mutable struct SameInstancesEvaluator <: AbstractEvaluator
+"""
 function evaluate(eval::SameInstancesEvaluator, variableHeuristic::AbstractVariableSelection, strategy::S; verbose::Bool=true) where{S<:SearchStrategy}
     for j in 1:eval.nbHeuristics
         heuristic = eval.metrics[1,j].heuristic
-        initsize = isa(heuristic, LearnedHeuristic) ? length(heuristic.agent.trajectory) : nothing
-
-        testmode!(heuristic, true)
-        for i in 1:eval.nbInstances
-            model = eval.instances[i]
-            reset_model!(model)
-
-            dt = @elapsed search!(model, strategy, variableHeuristic, heuristic)
-            eval.metrics[i,j](model,dt)
-
-            verbose && println(typeof(heuristic), " evaluated with: ", model.statistics.numberOfNodes, " nodes, taken ", dt, "s, number of solutions found : ", model.statistics.numberOfSolutions)
-        end 
-        testmode!(heuristic, false)
-        if !isnothing(initsize)
+        if isa(heuristic, LearnedHeuristic)             #LearnedHeuristic has to be evaluated at each evaluation step as long as the heuristic is updated along the training.
+            initsize = length(heuristic.agent.trajectory) 
+            testmode!(heuristic, true)
+            for i in 1:eval.nbInstances
+                model = eval.instances[i]
+                reset_model!(model)
+    
+                dt = @elapsed search!(model, strategy, variableHeuristic, heuristic)
+                eval.metrics[i,j](model,dt)
+                verbose && println(typeof(heuristic), " evaluated with: ", model.statistics.numberOfNodes, " nodes, taken ", dt, "s, number of solutions found : ", model.statistics.numberOfSolutions)
+            end 
+            testmode!(heuristic, false)
             @assert length(heuristic.agent.trajectory) == initsize "You have leaks in your evaluation pipeline!"
+            
+        
+        else                                            #BasicHeuristic only needs to be evaluated once, as it is not updated during training. As long as the eval has already been done once, it only repeat the search metrics for each evaluation step; 
+            for i in 1:eval.nbInstances
+                model = eval.instances[i]
+                reset_model!(model)
+                if eval.metrics[i,j].nbEpisodes == 0
+                dt = @elapsed search!(model, strategy, variableHeuristic, heuristic)
+                eval.metrics[i,j](model,dt)            
+                verbose && println(typeof(heuristic), " evaluated with: ", model.statistics.numberOfNodes, " nodes, taken ", dt, "s, number of solutions found : ", model.statistics.numberOfSolutions)
+                else
+                dt,numberOfNodes, numberOfSolutions = repeatlast!(eval.metrics[i,j])

```

Comment:
For every following evaluation step, we only repeat the metrics and avoid computing already existing results.